File: backend/src/keyboard_recommender/terminology/dictionary.py
"""
Curated community → internal-trait glosses.

**Structure**: each `CommunityTermDefinition` lists *relative* axis weights plus
`intrinsic_confidence`. The interpreter scales by match quality; overlapping terms *add*
unless you plug a different combiner later.

**NLP expansion**: add rows or load from JSON/YAML; optional `TerminologySource` can
emit soft distributions that get folded into the same `TraitContribution` tuples.
"""

from __future__ import annotations

from keyboard_recommender.terminology.models import CommunityTermDefinition, TraitContribution

Tc = TraitContribution


# Example mappings requested + common adjacent vocabulary (relative, signed shape).
COMMUNITY_TERMS: tuple[CommunityTermDefinition, ...] = (
    CommunityTermDefinition(
        id="thocky",
        canonical_label="thocky",
        synonyms=("thock", "thocky", "thocc", "도각도각", "도각", "저음도각", "묵직한소리"),
        trait_contributions=(
            Tc("deep_sound", 1.35),
            Tc("muted", 0.55),
            Tc("marbly", 0.85),
            Tc("high_pitch", -0.45),
            Tc("poppy", 0.25),
        ),
        intrinsic_confidence=0.82,
        notes="Low-mid resonance + rounded attack; overlaps with marbly/creamy.",
        tags=frozenset({"sound", "attack"}),
    ),
    CommunityTermDefinition(
        id="sound_deep",
        canonical_label="deep (sound)",
        synonyms=(
            "deep",
            "deeper",
            "deepest",
            "bassy",
            "subby",
            "low-end",
            "lowend",
            "저음",
            "저음성향",
            "깊은소리",
            "묵직한",
            "묵직한저음",
        ),
        trait_contributions=(
            Tc("deep_sound", 1.2),
            Tc("high_pitch", -0.35),
            Tc("marbly", 0.3),
            Tc("muted", 0.2),
        ),
        intrinsic_confidence=0.76,
        notes="Low-frequency / warm fundamental language; distinct from thocky meme bundle.",
        tags=frozenset({"sound"}),
    ),
    CommunityTermDefinition(
        id="sound_bright",
        canonical_label="bright / ping",
        synonyms=("bright", "brighter", "pingy", "ping", "shrill", "고음", "경쾌한고음", "쨍한", "맑은소리"),
        trait_contributions=(
            Tc("high_pitch", 1.05),
            Tc("poppy", 0.55),
            Tc("deep_sound", -0.4),
            Tc("muted", -0.35),
        ),
        intrinsic_confidence=0.7,
        notes="Upper-mid / attack-forward descriptors; overlaps clacky/poppy.",
        tags=frozenset({"sound", "attack"}),
    ),
    CommunityTermDefinition(
        id="sound_warm",
        canonical_label="warm (sound)",
        synonyms=("warm", "warmer", "warmth", "mellow", "웜톤", "따뜻한소리", "포근한소리"),
        trait_contributions=(
            Tc("deep_sound", 0.65),
            Tc("muted", 0.45),
            Tc("high_pitch", -0.35),
            Tc("smooth", 0.35),
        ),
        intrinsic_confidence=0.64,
        notes="Broad community term; softer than 'deep' alone on deep_sound.",
        tags=frozenset({"sound"}),
    ),
    CommunityTermDefinition(
        id="feel_tactile",
        canonical_label="tactile (feel)",
        synonyms=("tactile", "tactility", "bump", "bumpy", "snappy", "택타일", "구분감", "걸리는느낌"),
        trait_contributions=(
            Tc("strong_tactile", 1.15),
            Tc("firm_bottom_out", 0.35),
            Tc("smooth", -0.25),
        ),
        intrinsic_confidence=0.72,
        notes="Switch-feel wording; not identical to catalog 'tactile' tiers.",
        tags=frozenset({"feel"}),
    ),
    CommunityTermDefinition(
        id="feel_linear",
        canonical_label="linear (feel)",
        synonyms=("linear", "linears", "리니어", "매끈한", "직선적인"),
        trait_contributions=(
            Tc("smooth", 1.05),
            Tc("strong_tactile", -0.55),
            Tc("scratchy", -0.25),
        ),
        intrinsic_confidence=0.78,
        notes="No-bump preference language.",
        tags=frozenset({"feel"}),
    ),
    CommunityTermDefinition(
        id="feel_soft_bottom",
        canonical_label="soft bottom-out",
        synonyms=("mushy", "pillowy", "cushy", "cushioned", "푹신한", "쿠션감", "말랑한바닥"),
        trait_contributions=(
            Tc("soft_bottom_out", 1.1),
            Tc("firm_bottom_out", -0.45),
            Tc("muted", 0.25),
        ),
        intrinsic_confidence=0.66,
        notes="Bottom-out feel narrative; overlaps foamy/muted slightly.",
        tags=frozenset({"feel"}),
    ),
    CommunityTermDefinition(
        id="texture_scratchy",
        canonical_label="scratchy / rough",
        synonyms=("scratchy", "scratchier", "rough", "gritty", "raspy", "서걱서걱", "서걱", "거친질감"),
        trait_contributions=(
            Tc("scratchy", 1.15),
            Tc("smooth", -0.55),
            Tc("high_pitch", 0.2),
        ),
        intrinsic_confidence=0.62,
        notes="Surface / stem texture language; 'scratchy' surface form is the adjective row.",
        tags=frozenset({"texture"}),
    ),
    CommunityTermDefinition(
        id="creamy",
        canonical_label="creamy",
        synonyms=("creamy", "cream", "buttery", "크리미", "버터리", "쫀득한"),
        trait_contributions=(
            Tc("smooth", 0.95),
            Tc("deep_sound", 0.75),
            Tc("muted", 0.55),
            Tc("high_pitch", -0.35),
            Tc("scratchy", -0.4),
        ),
        intrinsic_confidence=0.68,
        notes="Smooth harmonic stack + softer treble; subjective vs thocky.",
        tags=frozenset({"sound", "texture"}),
    ),
    CommunityTermDefinition(
        id="clacky",
        canonical_label="clacky",
        synonyms=("clack", "clacky", "clack-forward", "클래키", "딱딱한소리", "경쾌한타건"),
        trait_contributions=(
            Tc("high_pitch", 1.2),
            Tc("poppy", 0.95),
            Tc("scratchy", 0.55),
            Tc("deep_sound", 0.15),
            Tc("muted", -0.65),
        ),
        intrinsic_confidence=0.8,
        notes="Bright, sharp front of hit; often anti-muted in community use.",
        tags=frozenset({"sound", "attack"}),
    ),
    CommunityTermDefinition(
        id="marbly",
        canonical_label="marbly",
        synonyms=("marble", "marbly", "마블리", "또각또각"),
        trait_contributions=(
            Tc("marbly", 1.25),
            Tc("deep_sound", 0.55),
            Tc("high_pitch", 0.15),
            Tc("poppy", 0.35),
        ),
        intrinsic_confidence=0.72,
        notes="Higher harmonic color riding on a deeper fundamental; overlaps thocky.",
        tags=frozenset({"sound", "harmonics"}),
    ),
    CommunityTermDefinition(
        id="poppy",
        canonical_label="poppy",
        synonyms=("poppy", "pop", "팝피", "통통튀는"),
        trait_contributions=(
            Tc("poppy", 1.15),
            Tc("high_pitch", 0.65),
            Tc("firm_bottom_out", 0.35),
            Tc("deep_sound", 0.2),
        ),
        intrinsic_confidence=0.74,
        notes="Short, snappy onset; can read as clacky-adjacent.",
        tags=frozenset({"sound", "attack"}),
    ),
    CommunityTermDefinition(
        id="muted",
        canonical_label="muted",
        synonyms=("muted", "mute", "quiet-sounding", "damped", "dampened", "뮤트", "조용한", "죽은소리", "차분한소리"),
        trait_contributions=(
            Tc("muted", 1.3),
            Tc("high_pitch", -0.75),
            Tc("soft_bottom_out", 0.35),
            Tc("scratchy", -0.45),
        ),
        intrinsic_confidence=0.78,
        notes="Damped treble / overall quieter presentation — not identical to 'quiet' dB.",
        tags=frozenset({"sound", "level"}),
    ),
    CommunityTermDefinition(
        id="foamy",
        canonical_label="foamy",
        synonyms=("foamy", "foam-heavy", "foam-maxed", "폼많은", "폼빵빵", "보글보글", "먹먹한"),
        trait_contributions=(
            Tc("muted", 0.85),
            Tc("soft_bottom_out", 0.55),
            Tc("high_pitch", -0.35),
            Tc("poppy", -0.25),
        ),
        intrinsic_confidence=0.62,
        notes="Case/foam damping narrative; correlates with muted/smooth, not a pure axis.",
        tags=frozenset({"setup", "damping"}),
    ),
    CommunityTermDefinition(
        id="case_hollow",
        canonical_label="hollow / case resonance",
        synonyms=("hollow", "echoey", "boomy", "통울림", "울림이큰", "텅빈소리", "부밍"),
        trait_contributions=(
            Tc("high_pitch", 0.45),
            Tc("muted", -0.55),
            Tc("deep_sound", 0.35),
            Tc("marbly", 0.25),
        ),
        intrinsic_confidence=0.64,
        notes="Case resonance / hollowness descriptors from community sound tests.",
        tags=frozenset({"sound", "resonance"}),
    ),
    CommunityTermDefinition(
        id="feel_bouncy",
        canonical_label="bouncy / chewy feel",
        synonyms=("bouncy", "chewy", "elastic", "쫀득쫀득", "쫀득함", "탱글한", "말랑한반발"),
        trait_contributions=(
            Tc("soft_bottom_out", 0.55),
            Tc("flexible", 0.75),
            Tc("firm_bottom_out", -0.35),
            Tc("smooth", 0.35),
        ),
        intrinsic_confidence=0.61,
        notes="Feel-forward terms often used for gasket/flex-heavy typing feel.",
        tags=frozenset({"feel", "mount"}),
    ),
    CommunityTermDefinition(
        id="feel_flexible",
        canonical_label="flexible typing feel",
        synonyms=("flex", "flexy", "gasket", "유연한", "플렉스", "말랑플렉스", "유격감"),
        trait_contributions=(
            Tc("flexible", 1.15),
            Tc("soft_bottom_out", 0.55),
            Tc("stiff", -0.75),
            Tc("firm_bottom_out", -0.25),
        ),
        intrinsic_confidence=0.7,
        notes="Mount/plate flexibility leaning descriptors.",
        tags=frozenset({"feel", "mount"}),
    ),
    CommunityTermDefinition(
        id="feel_stiff",
        canonical_label="stiff / rigid typing feel",
        synonyms=("stiff", "rigid", "hard", "빳빳한", "강성", "단단한", "단단한하우징"),
        trait_contributions=(
            Tc("stiff", 1.25),
            Tc("firm_bottom_out", 0.75),
            Tc("flexible", -0.85),
            Tc("soft_bottom_out", -0.45),
        ),
        intrinsic_confidence=0.72,
        notes="Rigid assembly / hard stop language.",
        tags=frozenset({"feel", "mount"}),
    ),
    CommunityTermDefinition(
        id="feel_light_force",
        canonical_label="light typing force",
        synonyms=("light", "feather", "easy-press", "가벼운", "저압", "살짝눌리는", "힘덜드는"),
        trait_contributions=(
            Tc("light_typing_force", 1.25),
            Tc("firm_bottom_out", -0.25),
            Tc("strong_tactile", -0.2),
            Tc("smooth", 0.2),
        ),
        intrinsic_confidence=0.66,
        notes="Low-force preference descriptors for long typing sessions.",
        tags=frozenset({"feel", "force"}),
    ),
    CommunityTermDefinition(
        id="feel_heavy_force",
        canonical_label="heavier typing force",
        synonyms=("heavy", "weighty", "hard-press", "무거운", "고압", "힘줘서누르는"),
        trait_contributions=(
            Tc("light_typing_force", -1.05),
            Tc("firm_bottom_out", 0.55),
            Tc("strong_tactile", 0.35),
            Tc("stiff", 0.35),
        ),
        intrinsic_confidence=0.63,
        notes="Heavier-force preference often paired with firmer feel.",
        tags=frozenset({"feel", "force"}),
    ),
    CommunityTermDefinition(
        id="silent",
        canonical_label="silent / low-noise",
        synonyms=("silent", "quiet", "low-noise", "저소음", "정숙한", "소음적은"),
        trait_contributions=(
            Tc("muted", 1.15),
            Tc("high_pitch", -0.65),
            Tc("scratchy", -0.35),
            Tc("soft_bottom_out", 0.25),
        ),
        intrinsic_confidence=0.74,
        notes="Explicit low-noise intent, separate from broad muted color terms.",
        tags=frozenset({"sound", "level"}),
    ),
    CommunityTermDefinition(
        id="crispy_attack",
        canonical_label="crisp / defined attack",
        synonyms=("crisp", "defined", "clean-attack", "또렷한", "선명한", "정갈한타건"),
        trait_contributions=(
            Tc("poppy", 0.95),
            Tc("high_pitch", 0.55),
            Tc("scratchy", 0.25),
            Tc("muted", -0.25),
        ),
        intrinsic_confidence=0.68,
        notes="Crisp attack language that sits between clacky and poppy.",
        tags=frozenset({"sound", "attack"}),
    ),
    CommunityTermDefinition(
        id="chalky_texture",
        canonical_label="chalky / dry texture",
        synonyms=("chalky", "dry", "powdery", "건조한", "분필같은", "드라이한질감"),
        trait_contributions=(
            Tc("scratchy", 0.95),
            Tc("smooth", -0.45),
            Tc("muted", 0.25),
            Tc("high_pitch", 0.2),
        ),
        intrinsic_confidence=0.58,
        notes="Dry/chalky descriptors mapped to texture roughness.",
        tags=frozenset({"texture"}),
    ),
)


def default_dictionary() -> tuple[CommunityTermDefinition, ...]:
    """Stable built-in lexicon; swap for DB-backed loader without changing the engine."""
    return COMMUNITY_TERMS

File: frontend/src/keyboard-terminology/dictionary.ts
/**
 * Community terminology → structured trait senses.
 *
 * **Multiple meanings:** entries can list several `TermSense` objects. Pick one with
 * strategies in `convert.ts`, or blend.
 *
 * **Editing:** add rows here; keep `traitBoost` sparse and on the same 0–10-ish scale as catalog items.
 */

import type { TermDictionaryEntry } from "@/keyboard-terminology/types";

/** Authoritative list — indexes are built in `indexes.ts`. */
export const TERMINOLOGY_DICTIONARY: TermDictionaryEntry[] = [
  {
    canonical: "thocky",
    aliases: ["thock", "thocc", "thockier", "도각도각", "도각", "도독도독", "저음도각", "묵직한소리", "묵직"],
    glossary: "묵직하고 낮은 톤의 바닥 타건감. 보통 깊은 하우징/플레이트 조합에서 잘 느껴집니다.",
    senses: [
      {
        id: "thocky:default",
        description: "Classic “thock” — depth-forward, less ping, rounded fundamental.",
        traitBoost: { deep_sound: 8, soft: 4, clacky: 2, smooth: 4 },
        confidence: 0.9,
      },
      {
        id: "thocky:plate_forward",
        description: "User emphasizes plate/case thock (still deep, slightly firmer presentation).",
        traitBoost: { deep_sound: 7, firm: 5, clacky: 3, smooth: 3 },
        confidence: 0.45,
      },
    ],
  },
  {
    canonical: "creamy",
    aliases: ["cream", "크리미", "버터리", "쫀득한", "쫀득쫀득", "부드러운"],
    glossary: "날카롭기보다 매끈하고 둥근 타건감. 윤활된 리니어 느낌을 말할 때 자주 쓰는 표현입니다.",
    senses: [
      {
        id: "creamy:default",
        description: "Smooth glide with softened, rounded feedback — often pairs with lubed linears.",
        traitBoost: { smooth: 8, soft: 6, deep_sound: 4, clacky: 2 },
        confidence: 0.85,
      },
    ],
  },
  {
    canonical: "clacky",
    aliases: ["clack", "clacky-forward", "클래키", "클락키", "경쾌한타건", "딱딱한소리", "쨍한소리"],
    glossary: "고음이 살아 있고 타건 어택이 또렷한 성향입니다.",
    senses: [
      {
        id: "clacky:default",
        description: "Sharp, present acoustics and firmer perceived impact.",
        traitBoost: { clacky: 9, firm: 5, deep_sound: 2, smooth: 3 },
        confidence: 0.9,
      },
    ],
  },
  {
    canonical: "marbly",
    aliases: ["marble", "마블리", "머블리", "또각또각", "마블톤"],
    glossary: "맥락에 따라 저음 마블리/고음 마블리로 나뉘는 표현입니다.",
    senses: [
      {
        id: "marbly:deep_smooth",
        description: "Dense, smooth fundamental with a stone-like roundness (often thock-leaning).",
        traitBoost: { deep_sound: 7, smooth: 6, soft: 4, clacky: 3 },
        confidence: 0.55,
      },
      {
        id: "marbly:bright_articulate",
        description: "Harder, more articulate “marble click” character — brighter edge.",
        traitBoost: { clacky: 7, firm: 6, smooth: 4, deep_sound: 4 },
        confidence: 0.55,
      },
    ],
  },
  {
    canonical: "muted",
    aliases: ["mute", "dampened", "뮤트", "조용한", "먹먹한", "차분한소리", "정숙한", "저소음"],
    glossary: "링/핑이 줄고 고음이 눌린 차분한 소리. 폼/테이프 세팅과 함께 자주 언급됩니다.",
    senses: [
      {
        id: "muted:default",
        description: "Overall quieter, more controlled harmonics; less clack-forward.",
        traitBoost: { soft: 8, clacky: 1, deep_sound: 4, smooth: 5 },
        confidence: 0.9,
      },
    ],
  },
  {
    canonical: "poppy",
    aliases: ["pop", "팝피", "파피", "통통튀는", "튀는느낌", "스냅감"],
    glossary: "통통 튀는 반발감과 또렷한 어택이 강조되는 성향입니다.",
    senses: [
      {
        id: "poppy:default",
        description: "Energetic, snappy feedback — often reads as brighter and more defined.",
        traitBoost: { clacky: 8, firm: 5, smooth: 4, tactile_strength: 4 },
        confidence: 0.75,
      },
      {
        id: "poppy:tactile_event",
        description: "Emphasis on a crisp tactile event (bump-forward wording in reviews).",
        traitBoost: { tactile_strength: 7, clacky: 6, firm: 4, smooth: 2 },
        confidence: 0.5,
      },
    ],
  },
  {
    canonical: "foamy",
    aliases: ["foamy", "폼많은", "보글보글", "폼빵빵"],
    glossary: "폼 세팅으로 소리가 정리되고 부드럽게 눌리는 느낌을 말할 때 쓰는 표현입니다.",
    senses: [
      {
        id: "foamy:default",
        description: "폼 감쇠가 많이 들어간 차분한 타건감.",
        traitBoost: { soft: 6, smooth: 4, deep_sound: 3, clacky: 1 },
        confidence: 0.72,
      },
    ],
  },
  {
    canonical: "scratchy",
    aliases: ["scratchy", "서걱서걱", "서걱", "거친질감", "사각사각", "까슬한"],
    glossary: "매끈하기보다 결이 느껴지는 거친 타건 질감입니다.",
    senses: [
      {
        id: "scratchy:default",
        description: "윤활이 덜 된 듯한 마찰감이 느껴지는 질감.",
        traitBoost: { smooth: -2, clacky: 5, firm: 4, tactile_strength: 4 },
        confidence: 0.67,
      },
    ],
  },
  {
    canonical: "deep",
    aliases: ["deep", "bassy", "lowpitch", "저음", "깊은소리", "묵직한저음"],
    glossary: "저음 중심의 묵직한 사운드 성향입니다.",
    senses: [
      {
        id: "deep:default",
        description: "고음보다 저음 중심으로 무게감이 느껴지는 소리.",
        traitBoost: { deep_sound: 8, clacky: 1, soft: 3 },
        confidence: 0.82,
      },
    ],
  },
  {
    canonical: "bright",
    aliases: ["bright", "pingy", "고음", "쨍한", "맑은소리", "경쾌한고음"],
    glossary: "고음이 살아 있고 어택이 선명한 성향입니다.",
    senses: [
      {
        id: "bright:default",
        description: "상단 고음이 또렷하고 반응이 빠른 느낌.",
        traitBoost: { clacky: 8, firm: 3, deep_sound: 1 },
        confidence: 0.79,
      },
    ],
  },
  {
    canonical: "tactile",
    aliases: ["tactile", "택타일", "구분감", "범프", "걸리는느낌"],
    glossary: "눌렀을 때 구분감(범프)이 느껴지는 성향입니다.",
    senses: [
      {
        id: "tactile:default",
        description: "명확한 구분감이 있는 입력 피드백.",
        traitBoost: { tactile_strength: 8, firm: 3, smooth: -1 },
        confidence: 0.83,
      },
    ],
  },
  {
    canonical: "linear",
    aliases: ["linear", "리니어", "매끈한", "미끄러지는"],
    glossary: "구분감보다 매끈한 직선형 입력 감각입니다.",
    senses: [
      {
        id: "linear:default",
        description: "범프가 적고 부드럽게 이어지는 입력감.",
        traitBoost: { smooth: 8, tactile_strength: -2, soft: 2 },
        confidence: 0.84,
      },
    ],
  },
  {
    canonical: "bouncy",
    aliases: ["bouncy", "탄성있는", "탱글한", "쫀득쫀득", "말랑한반발"],
    glossary: "탄성감이 있고 말랑하게 반발하는 타건감입니다.",
    senses: [
      {
        id: "bouncy:default",
        description: "부드럽고 탄성 있는 반발감 중심.",
        traitBoost: { soft: 6, smooth: 4, firm: -1, tactile_strength: 2 },
        confidence: 0.73,
      },
    ],
  },
  {
    canonical: "stiff",
    aliases: ["stiff", "rigid", "강성", "빳빳한", "단단한하우징", "단단한"],
    glossary: "유연함보다 단단하고 고정된 타건감을 원하는 표현입니다.",
    senses: [
      {
        id: "stiff:default",
        description: "하드한 바닥감과 단단한 입력 피드백.",
        traitBoost: { firm: 8, soft: -2, clacky: 3 },
        confidence: 0.76,
      },
    ],
  },
  {
    canonical: "light-force",
    aliases: ["light", "저압", "가벼운", "살짝눌리는", "힘덜드는"],
    glossary: "적은 힘으로 가볍게 눌리는 입력감을 뜻합니다.",
    senses: [
      {
        id: "light-force:default",
        description: "부담 없는 키압 중심의 장시간 타이핑 성향.",
        traitBoost: { soft: 3, smooth: 3, firm: -1, tactile_strength: -1 },
        confidence: 0.68,
      },
    ],
  },
  {
    canonical: "hollow",
    aliases: ["hollow", "통울림", "텅빈소리", "울림이큰", "부밍"],
    glossary: "하우징 내부 공진이 느껴지는 울림 성향입니다.",
    senses: [
      {
        id: "hollow:default",
        description: "울림이 강조되어 차분함이 줄어드는 케이스 공진 느낌.",
        traitBoost: { deep_sound: 3, clacky: 3, soft: -2, firm: 2 },
        confidence: 0.65,
      },
    ],
  },
];
